### PR TITLE
Ee 16843 PassStatisticsAccumulator

### PR DESCRIPTION
**Pull request by owainrobertsbjss**
A class which when passed lists of AuditResultsByNino will store the best result for every Nino. When the result() method is called, it will filter all the results in range and return the statistics. I intend to use this class in a service (to be written) that gets results for each day from the audit service and passes to this class.

The code is isolated and should not change any function of the service.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+import static java.util.Objects.isNull;
+
+class PassStatisticsAccumulator {
+
+    private final LocalDate fromDate;
+    private final LocalDate toDate;
+
+    private final Map<String, BestResult> bestResultByNino;
+    private final AuditResultTypeComparator resultTypeComparator;
```
Comment (Tim-Hammonds) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java]: Comparator seems overkill unless you ever have a collection of the compared type.

File: src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    PassStatisticsAccumulator(LocalDate fromDate, LocalDate toDate) {
+        this.fromDate = fromDate;
+        this.toDate = toDate;
+        bestResultByNino = new HashMap<>();
+        resultTypeComparator = new AuditResultTypeComparator();
+    }
+
+    void accumulate(List<AuditResultByNino> records) {
+        for (AuditResultByNino record : records) {
```
Comment (Tim-Hammonds) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java]: Maybe stream?

File: src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            BestResult currentBestResult = bestResultByNino.get(record.nino());
+
+            if (isNull(currentBestResult)
```
Comment (Tim-Hammonds) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java]: All this condition can be in the Comparator, or in the comparison function

File: src/test/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulatorMultipleCallsTest.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    public void accumulate_firstPassTooEarlySecondInRange_onlyCountOne() {
+        accumulator.accumulate(singleResult(FROM_DATE.minusDays(1), PASS));
+        accumulator.accumulate(singleResult(FROM_DATE, PASS));
+
+        assertThat(accumulator.result())
+            .isEqualTo(statisticsForCounts(1, 0, 0, 0));
+    }
+
+    @Test
+    public void accumulate_firstPassInRangeSecondTooLate_notCounted() {
```
Comment (Tim-Hammonds) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulatorMultipleCallsTest.java]: You'll have to explain the rules for me to understand this test

Comment (owainrobertsbjss) [reply src/test/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulatorMultipleCallsTest.java]: The rules:

- An AuditResult is in the date range if its best result between the dates (inclusive). 
- The ranking of results is PASS > FAIL > NOTFOUND > ERROR
- Where we have multiple records with the same result, we take the most recent.

So in this case we have 2 PASS but as the latest is after the TO_DATE, we don't count the results for this Nino towards the stats.

File: src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        long passes = countsByResult.getOrDefault(AuditResultType.PASS, 0L);
+        long failures = countsByResult.getOrDefault(AuditResultType.FAIL, 0L);
+        long notFound = countsByResult.getOrDefault(AuditResultType.NOTFOUND, 0L);
+        long errors = countsByResult.getOrDefault(AuditResultType.ERROR, 0L);
+
+        return new PassRateStatistics(fromDate, toDate, totalRequests, passes, failures, notFound, errors);
+    }
+
+    private boolean betterThanCurrentBest(AuditResultByNino record, BestResult currentBestResult) {
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/statistics/PassStatisticsAccumulator.java]: The intention was for the AuditResultConsolidator to calculate the best / newest result per nino, so you shouldn't need to do that here if the data has already been "consolidated".  That said, the method names in the consolidator don't make that very clear, so maybe you could do some refactoring in the consolidator to make its usage clear.

**Review by mikehalmaho — CHANGES_REQUESTED** [review]

**Review by mikehalmaho — APPROVED** [review]

**Pull request merged**